src/dataflow-types/src/client/replicated.rs

```diff
@@ -25,18 +25,77 @@
 use std::collections::{HashMap, HashSet};
 
 use timely::progress::{frontier::MutableAntichain, Antichain};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
 
 use crate::client::Peek;
 use mz_repr::GlobalId;
 
 use super::{ComputeClient, GenericClient};
 use super::{ComputeCommand, ComputeResponse};
 
+// `Client` must have a cancelation-safe `recv` method.
+pub fn spawn_client_tasklet<
+    C: Send + 'static,
+    R: Send + 'static,
+    Client: GenericClient<C, R> + 'static,
+    Name: AsRef<str>,
+    NameClosure: FnOnce() -> Name,
+>(
+    mut client: Client,
+    nc: NameClosure,
+) -> (
+    JoinHandle<()>,
+    UnboundedSender<C>,
+    UnboundedReceiver<Result<R, anyhow::Error>>,
+) {
+    let (cmd_tx, mut cmd_rx) = unbounded_channel();
+    let (response_tx, response_rx) = unbounded_channel();
+    let handle = mz_ore::task::spawn(nc, async move {
+        loop {
+            tokio::select! {
+                m = cmd_rx.recv() => {
+                    match m {
+                        Some(c) => {
+                            // Issues should be detected, and
+                            // reconnect attempted, on the `recv` path.
+                            let _ = client.send(c).await;
+                        },
+                        None => break,
+                    }
+                },
+                m = client.recv() => {
```

Comment:
isn't this still the same bug? this `recv` future is not cancel-safe right?

Reply:
The _inner_ client's `recv` is cancellation safe.

Reply:
Because this is generic, we should document this requirement on this function!

It looks like when I follow this down, its instantiated with a `RemoteClient` that wraps a `Partitioned`, whose `recv` looks like its subtle to reason about:

https://github.com/MaterializeInc/materialize/blob/main/src/dataflow-types/src/client/partitioned.rs#L82-L110 could be cancelled before receiving all the parts of a response, but it should not drop that data as it stores it in its `state`, as long as its underlying client is cancel-safe, which `TcpClient` appears to be? This is a fairly complex stack to reason about

It may worth making the control flow uglier here by ensuring we always poll the client recv, but because the `send` and `recv` both take `&mut self` and are on the same client, its hard, perhaps its time to split up the `GenericClient` trait into 2 traits, though that would be a lot of work

Reply:
> It may worth making the control flow uglier here by ensuring we always poll the client recv, but because the send and recv both take &mut self and are on the same client, its hard, perhaps its time to split up the GenericClient trait into 2 traits, though that would be a lot of work

Yeah, the problem is that they both _need_ exclusive access to the same state. If you split it into two halves, then you need all the client state to go behind an `Arc<Mutex<State>>`.